src/ComplexMarkdownParser.php

```diff
@@ -0,0 +1,168 @@
+<?php
+
+namespace Spatie\YamlFrontMatter;
+
+use Symfony\Component\Yaml\Yaml;
+
+/**
+ * A parser that can handle Markdown that contains Markdown.
+ *
+ * Attempts to follow the practices defined in https://jekyllrb.com/docs/front-matter/.
+ */
+class ComplexMarkdownParser
+{
+    /**
+     * The Markdown content.
+     * @var string
+     */
+    protected $content;
+
+    /**
+     * The document string as an array of lines,
+     * making it easier to understand and work with.
+     * @var array
+     */
+    protected $lines;
+
+    /**
+     * The line number of the starting front matter control block.
+     * @var int|null
+     */
+    public $frontMatterStartLine;
+
+    /**
+     * The line number of the ending front matter control block.
+     * @var int|null
+     */
+    public $frontMatterEndLine;
+
+    /**
+     * Construct the parser.
+     * @param string $content The Markdown content to parse.
+     */
+    public function __construct(string $content)
+    {
+        $this->content = $content;
+        $this->lines = explode("\n", $this->content);
+    }
+
+    /**
+     * Parse the Markdown content.
+     * @return Document
+     */
+    public function parse(): Document
+    {
+        $this->findFrontMatterStartAndEndLineNumbers();
+
+        if (!$this->hasFrontMatter()) {
+            return new Document([], $this->content);
+        }
+
+        $matter = $this->getFrontMatter();
+        $body = $this->getBody();
+
+        $matter = Yaml::parse($matter);
+
+        return new Document($matter, $body);
+    }
+
+    /**
+     * Is a given line a front matter control block?
+     *
+     * A control block is a line that starts with three dashes.
+     * @param string $line
+     * @return bool
+     */
+    private function isFrontMatterControlBlock(string $line): bool
```

Comment:
We use `protected` by default instead of `private`. Change all uses of `private` to `protected`.

Additionally, this helper function is only being used only later in the code. Move this function underneath the code where it is being used.